File: admin/ios/TigerAdmin/Sources/Models/DomainModels.swift
//
//  DomainModels.swift
//  TigerAdmin
//
//  Generic Codable wrapper for the 12 admin domain endpoints
//  (futures, options, copy-trading, convert, onramp, offramp, p2p-clients,
//  p2p-merchants, partners, rewards, marketing, roles, permissions).
//

import Foundation

// Loose wrapper: fields vary per domain, so all fields except id are optional.
// Decodes defensively (any numeric id is coerced to String) so a missing or
// differently-typed field never breaks the whole list.
struct DomainRecord: Codable, Identifiable {
    let id: String
    var name: String?
    var symbol: String?
    var status: String?
    var email: String?
    var type: String?
    var description: String?
    var resource: String?
    var action: String?
    var permissions: [String]?
    var trader: String?
    var followers: Int?
    var leverage: String?
    var margin: String?
    var strike: String?
    var expiry: String?
    var fromAsset: String?
    var toAsset: String?
    var amount: String?
    var rate: String?
    var user: String?
    var provider: String?
    var verified: Bool?
    var campaign: String?

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: AnyKey.self)
        func str(_ k: String) -> String? {
            if let v = try? c.decode(String.self, forKey: AnyKey(stringValue: k)) { return v }
            if let v = try? c.decode(Int.self, forKey: AnyKey(stringValue: k)) { return String(v) }
            if let v = try? c.decode(Double.self, forKey: AnyKey(stringValue: k)) { return String(v) }
            return nil
        }
        func bool(_ k: String) -> Bool? { try? c.decode(Bool.self, forKey: AnyKey(stringValue: k)) }
        func ints(_ k: String) -> [String]? { try? c.decode([String].self, forKey: AnyKey(stringValue: k)) }
        func intv(_ k: String) -> Int? { try? c.decode(Int.self, forKey: AnyKey(stringValue: k)) }
        id = str("id") ?? UUID().uuidString
        name = str("name"); symbol = str("symbol"); status = str("status")
        email = str("email"); type = str("type"); description = str("description")
        resource = str("resource"); action = str("action")
        permissions = ints("permissions")
        trader = str("trader"); followers = intv("followers")
        leverage = str("leverage"); margin = str("margin")
        strike = str("strike"); expiry = str("expiry")
        fromAsset = str("from_asset"); toAsset = str("to_asset")
        amount = str("amount"); rate = str("rate")
        user = str("user"); provider = str("provider")
        verified = bool("verified"); campaign = str("campaign")
    }

    func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: AnyKey.self)
        try c.encode(id, forKey: AnyKey(stringValue: "id"))
        try c.encodeIfPresent(name, forKey: AnyKey(stringValue: "name"))
        try c.encodeIfPresent(status, forKey: AnyKey(stringValue: "status"))
    }
}

// Coding key that accepts arbitrary string keys (used by DomainRecord).
struct AnyKey: CodingKey {
    var stringValue: String
    init(stringValue: String) { self.stringValue = stringValue }
    init?(stringValue: String) { self.stringValue = stringValue }
    var intValue: Int? { nil }
    init?(intValue: Int) { nil }
}

// MARK: - New admin domains (bots, bots-clients, project-teams, liquidity-sources)
// Loose Codable structs mirroring the admin/go (port 9093) payloads. Fields are
// optional and decoded defensively so a partial/missing field never breaks a list.

struct BotDomainRecord: Codable, Identifiable {
    let id: String
    var name: String?
    var botType: String?
    var leverage: Int?
    var status: String?
    var createdAt: String?
    var updatedAt: String?

    enum CodingKeys: String, CodingKey {
        case id, name, status
        case botType = "bot_type"
        case leverage
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        func str(_ k: KeyedDecodingContainer<CodingKeys>.Key) -> String? {
            if let v = try? c.decode(String.self, forKey: k) { return v }
            if let v = try? c.decode(Int.self, forKey: k) { return String(v) }
            return nil
        }
        id = str(.id) ?? UUID().uuidString
        name = str(.name); botType = str(.botType); status = str(.status)
        leverage = try? c.decode(Int.self, forKey: .leverage)
        createdAt = str(.createdAt); updatedAt = str(.updatedAt)
    }

    func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        try c.encode(id, forKey: .id)
        try c.encodeIfPresent(name, forKey: .name)
        try c.encodeIfPresent(botType, forKey: .botType)
        try c.encodeIfPresent(leverage, forKey: .leverage)
        try c.encodeIfPresent(status, forKey: .status)
    }
}

struct BotDomainRequest: Codable {
    var name: String
    var botType: String?
    var leverage: Int?
    enum CodingKeys: String, CodingKey { case name; case botType = "bot_type"; case leverage }
}

struct BotTierRecord: Codable, Identifiable {
    let id: String
    var name: String?
    var minVolume: String?
    var maxVolume: String?
    var feeRate: String?
    enum CodingKeys: String, CodingKey {
        case id, name
        case minVolume = "min_volume"; case maxVolume = "max_volume"; case feeRate = "fee_rate"
    }
}

struct BotTierRequest: Codable {
    var name: String?
    var minVolume: String?
    var maxVolume: String?
    var feeRate: String?
    enum CodingKeys: String, CodingKey {
        case name; case minVolume = "min_volume"; case maxVolume = "max_volume"; case feeRate = "fee_rate"
    }
}

struct BotsClientRecord: Codable, Identifiable {
    let id: String
    var name: String?
    var clientId: String?
    var apiKey: String?
    var status: String?
    var createdAt: String?
    enum CodingKeys: String, CodingKey {
        case id, name, status
        case clientId = "client_id"; case apiKey = "api_key"; case createdAt = "created_at"
    }
}

struct BotsClientRequest: Codable {
    var name: String
    var clientId: String?
    var apiKey: String?
    enum CodingKeys: String, CodingKey { case name; case clientId = "client_id"; case apiKey = "api_key" }
}

struct ProjectTeamRecord: Codable, Identifiable {
    let id: String
    var name: String?
    var description: String?
    var status: String?
    var createdAt: String?
    enum CodingKeys: String, CodingKey { case id, name, description, status; case createdAt = "created_at" }
}

struct ProjectTeamRequest: Codable {
    var name: String
    var description: String?
}

struct ProjectTeamMemberRecord: Codable, Identifiable {
    let id: String
    var teamId: String?
    var userId: String?
    var role: String?
    var status: String?
    enum CodingKeys: String, CodingKey {
        case id, role, status
        case teamId = "team_id"; case userId = "user_id"
    }
}

struct AddProjectTeamMemberRequest: Codable {
    var userId: String
    var role: String?
    enum CodingKeys: String, CodingKey { case userId = "user_id"; case role }
}

struct LiquiditySourceRecord: Codable, Identifiable {
    let id: String
    var name: String?
    var provider: String?
    var priority: Int?
    var status: String?
    var healthy: Bool?
    var lastCheck: String?
    enum CodingKeys: String, CodingKey {
        case id, name, provider, priority, status
        case healthy; case lastCheck = "last_check"
    }
}

struct LiquiditySourceRequest: Codable {
    var name: String
    var provider: String?
    var priority: Int?
}

struct SetLiquiditySourcePriorityRequest: Codable {
    let priority: Int
}

struct DomainStats: Codable {
    var total: Int?
    var active: Int?
    var inactive: Int?
}

File: admin/android/app/src/main/java/com/tigerwallet/admin/data/model/DomainModels.kt
package com.tigerwallet.admin.data.model

import com.google.gson.annotations.SerializedName

/**
 * Domain models for the 12 admin families driven by the admin/go backend (port 9093).
 * These back the futures / options / copy-trading / convert / onramp / offramp /
 * p2p-clients / p2p-merchants / partners / rewards / marketing / roles admin screens.
 */

data class StatusUpdateRequest(
    @SerializedName("status") val status: String
)

data class RejectRequest(
    @SerializedName("reason") val reason: String
)

data class AssignRoleRequest(
    @SerializedName("roleId") val roleId: String
)

data class FuturesRecord(
    @SerializedName("id") val id: String,
    @SerializedName("symbol") val symbol: String? = null,
    @SerializedName("name") val name: String? = null,
    @SerializedName("status") val status: String,
    @SerializedName("leverage") val leverage: Int? = null,
    @SerializedName("margin") val margin: String? = null,
    @SerializedName("created_at") val createdAt: String? = null,
    @SerializedName("updated_at") val updatedAt: String? = null
)

data class FuturesRequest(
    @SerializedName("symbol") val symbol: String,
    @SerializedName("name") val name: String,
    @SerializedName("leverage") val leverage: Int? = null,
    @SerializedName("margin") val margin: String? = null
)

data class OptionsRecord(
    @SerializedName("id") val id: String,
    @SerializedName("symbol") val symbol: String? = null,
    @SerializedName("name") val name: String? = null,
    @SerializedName("status") val status: String,
    @SerializedName("strike") val strike: String? = null,
    @SerializedName("expiry") val expiry: String? = null,
    @SerializedName("created_at") val createdAt: String? = null,
    @SerializedName("updated_at") val updatedAt: String? = null
)

data class OptionsRequest(
    @SerializedName("symbol") val symbol: String,
    @SerializedName("name") val name: String,
    @SerializedName("strike") val strike: String? = null,
    @SerializedName("expiry") val expiry: String? = null
)

data class CopyTradingRecord(
    @SerializedName("id") val id: String,
    @SerializedName("name") val name: String? = null,
    @SerializedName("status") val status: String,
    @SerializedName("trader") val trader: String? = null,
    @SerializedName("followers") val followers: Int? = null,
    @SerializedName("created_at") val createdAt: String? = null,
    @SerializedName("updated_at") val updatedAt: String? = null
)

data class CopyTradingRequest(
    @SerializedName("name") val name: String,
    @SerializedName("trader") val trader: String? = null
)

data class ConvertRecord(
    @SerializedName("id") val id: String,
    @SerializedName("from_asset") val fromAsset: String? = null,
    @SerializedName("to_asset") val toAsset: String? = null,
    @SerializedName("status") val status: String,
    @SerializedName("amount") val amount: String? = null,
    @SerializedName("rate") val rate: String? = null,
    @SerializedName("created_at") val createdAt: String? = null,
    @SerializedName("updated_at") val updatedAt: String? = null
)

data class ConvertRequest(
    @SerializedName("from_asset") val fromAsset: String,
    @SerializedName("to_asset") val toAsset: String,
    @SerializedName("amount") val amount: String? = null,
    @SerializedName("rate") val rate: String? = null
)

data class OnRampRecord(
    @SerializedName("id") val id: String,
    @SerializedName("user") val user: String? = null,
    @SerializedName("asset") val asset: String? = null,
    @SerializedName("amount") val amount: String? = null,
    @SerializedName("status") val status: String,
    @SerializedName("provider") val provider: String? = null,
    @SerializedName("created_at") val createdAt: String? = null,
    @SerializedName("updated_at") val updatedAt: String? = null
)

data class OnRampRequest(
    @SerializedName("asset") val asset: String,
    @SerializedName("amount") val amount: String,
    @SerializedName("provider") val provider: String? = null
)

data class OffRampRecord(
    @SerializedName("id") val id: String,
    @SerializedName("user") val user: String? = null,
    @SerializedName("asset") val asset: String? = null,
    @SerializedName("amount") val amount: String? = null,
    @SerializedName("status") val status: String,
    @SerializedName("provider") val provider: String? = null,
    @SerializedName("created_at") val createdAt: String? = null,
    @SerializedName("updated_at") val updatedAt: String? = null
)

data class OffRampRequest(
    @SerializedName("asset") val asset: String,
    @SerializedName("amount") val amount: String,
    @SerializedName("provider") val provider: String? = null
)

data class P2PClientRecord(
    @SerializedName("id") val id: String,
    @SerializedName("name") val name: String? = null,
    @SerializedName("email") val email: String? = null,
    @SerializedName("status") val status: String,
    @SerializedName("created_at") val createdAt: String? = null,
    @SerializedName("updated_at") val updatedAt: String? = null
)

data class P2PClientRequest(
    @SerializedName("name") val name: String,
    @SerializedName("email") val email: String? = null
)

data class P2PMerchantRecord(
    @SerializedName("id") val id: String,
    @SerializedName("name") val name: String? = null,
    @SerializedName("email") val email: String? = null,
    @SerializedName("status") val status: String,
    @SerializedName("verified") val verified: Boolean? = null,
    @SerializedName("created_at") val createdAt: String? = null,
    @SerializedName("updated_at") val updatedAt: String? = null
)

data class P2PMerchantRequest(
    @SerializedName("name") val name: String,
    @SerializedName("email") val email: String? = null
)

data class PartnerRecord(
    @SerializedName("id") val id: String,
    @SerializedName("name") val name: String? = null,
    @SerializedName("type") val type: String? = null,
    @SerializedName("status") val status: String,
    @SerializedName("approved") val approved: Boolean? = null,
    @SerializedName("created_at") val createdAt: String? = null,
    @SerializedName("updated_at") val updatedAt: String? = null
)

data class PartnerRequest(
    @SerializedName("name") val name: String,
    @SerializedName("type") val type: String? = null
)

data class RewardRecord(
    @SerializedName("id") val id: String,
    @SerializedName("name") val name: String? = null,
    @SerializedName("type") val type: String? = null,
    @SerializedName("status") val status: String,
    @SerializedName("amount") val amount: String? = null,
    @SerializedName("created_at") val createdAt: String? = null,
    @SerializedName("updated_at") val updatedAt: String? = null
)

data class RewardRequest(
    @SerializedName("name") val name: String,
    @SerializedName("type") val type: String? = null,
    @SerializedName("amount") val amount: String? = null
)

data class MarketingRecord(
    @SerializedName("id") val id: String,
    @SerializedName("name") val name: String? = null,
    @SerializedName("campaign") val campaign: String? = null,
    @SerializedName("status") val status: String,
    @SerializedName("created_at") val createdAt: String? = null,
    @SerializedName("updated_at") val updatedAt: String? = null
)

data class MarketingRequest(
    @SerializedName("name") val name: String,
    @SerializedName("campaign") val campaign: String? = null
)

data class RoleRecord(
    @SerializedName("id") val id: String,
    @SerializedName("name") val name: String,
    @SerializedName("description") val description: String? = null,
    @SerializedName("permissions") val permissions: List<String>? = null,
    @SerializedName("created_at") val createdAt: String? = null,
    @SerializedName("updated_at") val updatedAt: String? = null
)

data class RoleRequest(
    @SerializedName("name") val name: String,
    @SerializedName("description") val description: String? = null,
    @SerializedName("permissions") val permissions: List<String>? = null
)

data class PermissionRecord(
    @SerializedName("id") val id: String,
    @SerializedName("name") val name: String,
    @SerializedName("description") val description: String? = null,
    @SerializedName("resource") val resource: String? = null,
    @SerializedName("action") val action: String? = null,
    @SerializedName("created_at") val createdAt: String? = null
)

data class PermissionRequest(
    @SerializedName("name") val name: String,
    @SerializedName("description") val description: String? = null,
    @SerializedName("resource") val resource: String? = null,
    @SerializedName("action") val action: String? = null
)

// ============================================================================
// New admin domains (bots, bots-clients, project-teams, liquidity-sources)
// backed by the admin/go service on port 9093. Each domain exposes CRUD +
// setStatus, plus the domain-specific sub-resources documented in
// admin/cpp/include/admin_bots.hpp.
// ============================================================================

// ---------- Bots ----------
data class BotRecord(
    @SerializedName("id") val id: String,
    @SerializedName("name") val name: String? = null,
    @SerializedName("status") val status: String,
    @SerializedName("bot_type") val botType: String? = null,
    @SerializedName("user_id") val userId: String? = null,
    @SerializedName("leverage") val leverage: Int? = null,
    @SerializedName("created_at") val createdAt: String? = null,
    @SerializedName("updated_at") val updatedAt: String? = null
)

data class BotRequest(
    @SerializedName("name") val name: String,
    @SerializedName("bot_type") val botType: String? = null,
    @SerializedName("leverage") val leverage: Int? = null
)

data class BotTierRecord(
    @SerializedName("id") val id: String,
    @SerializedName("name") val name: String? = null,
    @SerializedName("level") val level: Int? = null,
    @SerializedName("status") val status: String? = null,
    @SerializedName("created_at") val createdAt: String? = null,
    @SerializedName("updated_at") val updatedAt: String? = null
)

data class BotTierRequest(
    @SerializedName("name") val name: String,
    @SerializedName("level") val level: Int? = null,
    @SerializedName("status") val status: String? = null
)

// ---------- Bots Clients ----------
data class BotsClientRecord(
    @SerializedName("id") val id: String,
    @SerializedName("name") val name: String? = null,
    @SerializedName("email") val email: String? = null,
    @SerializedName("status") val status: String,
    @SerializedName("client_id") val clientId: String? = null,
    @SerializedName("created_at") val createdAt: String? = null,
    @SerializedName("updated_at") val updatedAt: String? = null
)

data class BotsClientRequest(
    @SerializedName("name") val name: String,
    @SerializedName("email") val email: String? = null,
    @SerializedName("client_id") val clientId: String? = null
)

// ---------- Project Teams ----------
data class ProjectTeamRecord(
    @SerializedName("id") val id: String,
    @SerializedName("name") val name: String? = null,
    @SerializedName("status") val status: String,
    @SerializedName("description") val description: String? = null,
    @SerializedName("created_at") val createdAt: String? = null,
    @SerializedName("updated_at") val updatedAt: String? = null
)

data class ProjectTeamRequest(
    @SerializedName("name") val name: String,
    @SerializedName("description") val description: String? = null
)

data class ProjectTeamMemberRecord(
    @SerializedName("id") val id: String,
    @SerializedName("team_id") val teamId: String? = null,
    @SerializedName("user_id") val userId: String? = null,
    @SerializedName("email") val email: String? = null,
    @SerializedName("role") val role: String? = null,
    @SerializedName("created_at") val createdAt: String? = null
)

data class AddProjectTeamMemberRequest(
    @SerializedName("user_id") val userId: String,
    @SerializedName("role") val role: String? = null
)

// ---------- Liquidity Sources ----------
data class LiquiditySourceRecord(
    @SerializedName("id") val id: String,
    @SerializedName("name") val name: String? = null,
    @SerializedName("status") val status: String,
    @SerializedName("type") val type: String? = null,
    @SerializedName("priority") val priority: Int? = null,
    @SerializedName("chain") val chain: String? = null,
    @SerializedName("created_at") val createdAt: String? = null,
    @SerializedName("updated_at") val updatedAt: String? = null
)

data class LiquiditySourceRequest(
    @SerializedName("name") val name: String,
    @SerializedName("type") val type: String? = null,
    @SerializedName("chain") val chain: String? = null,
    @SerializedName("priority") val priority: Int? = null
)

data class SetLiquiditySourcePriorityRequest(
    @SerializedName("priority") val priority: Int
)

// Generic stats wrapper: stats payloads vary per domain, so decode loosely.
data class DomainStatsResponse(
    @SerializedName("total") val total: Long? = null,
    @SerializedName("active") val active: Long? = null,
    @SerializedName("inactive") val inactive: Long? = null,
    @SerializedName("healthy") val healthy: Long? = null,
    @SerializedName("unhealthy") val unhealthy: Long? = null,
    @SerializedName("total_volume") val totalVolume: String? = null,
    @SerializedName("total_pnl") val totalPnl: String? = null
)

